File: bench/deepseek-vision-compare/metrics.mjs
// SPDX-License-Identifier: AGPL-3.0-only
// Same equations as the pinned compare_reference.py:36-50; no teacher execution.
export const THRESHOLDS = Object.freeze({ cosine_min: 0.999, worst_row_min: 0.995,
  relative_l2_max: 0.05 });

const FIELDS = ['cosine', 'worst_row_cosine', 'relative_l2', 'max_abs_error', 'exact_fraction'];

// Compensated FP64 accumulation avoids a long serial reduction losing small
// terms. CUDA's parallel FP64 reduction can still differ in its last bits.
class Sum {
  value = 0;
  correction = 0;
  add(value) {
    const next = this.value + value;
    this.correction += Math.abs(this.value) >= Math.abs(value)
      ? (this.value - next) + value : (value - next) + this.value;
    this.value = next;
  }
  total() { return this.value + this.correction; }
}

export function compareBf16(actual, reference, rows, cols) {
  if (!Number.isSafeInteger(rows) || rows < 1 || rows > 384
    || !Number.isSafeInteger(cols) || cols < 1 || cols > 4096
    || !Buffer.isBuffer(actual) || !Buffer.isBuffer(reference)
    || actual.length !== rows * cols * 2 || reference.length !== actual.length) {
    throw new Error('invalid BF16 comparison geometry or exact byte extent');
  }
  const word = new DataView(new ArrayBuffer(4));
  const decode = (buffer, index) => {
    word.setUint32(0, buffer.readUInt16LE(index * 2) * 65536, true);
    const value = word.getFloat32(0, true);
    if (!Number.isFinite(value)) throw new Error('nonfinite BF16 comparison value');
    return value;
  };
  const a2 = new Sum(), b2 = new Sum(), dot = new Sum(), difference2 = new Sum();
  let worst = Infinity, maxError = 0, equal = 0;
  for (let row = 0; row < rows; row++) {
    const ra2 = new Sum(), rb2 = new Sum(), rdot = new Sum();
    for (let col = 0; col < cols; col++) {
      const index = row * cols + col;
      const a = decode(actual, index), b = decode(reference, index);
      ra2.add(a * a); rb2.add(b * b); rdot.add(a * b);
      const difference = a - b;
      difference2.add(difference * difference);
      maxError = Math.max(maxError, Math.abs(difference));
      if (a === b) equal++; // Numeric equality, including +0 versus -0, as Torch.
    }
    const aa = ra2.total(), bb = rb2.total(), ab = rdot.total();
    if (!(aa > 0) || !(bb > 0)) throw new Error('zero norm comparison row');
    worst = Math.min(worst, ab / (Math.sqrt(aa) * Math.sqrt(bb)));
    a2.add(aa); b2.add(bb); dot.add(ab);
  }
  const referenceNorm = Math.sqrt(b2.total());
  const result = {
    cosine: dot.total() / (Math.sqrt(a2.total()) * referenceNorm),
    worst_row_cosine: worst,
    relative_l2: Math.sqrt(difference2.total()) / referenceNorm,
    max_abs_error: maxError,
    exact_fraction: Math.fround(equal / (rows * cols)),
  };
  if (!FIELDS.every(key => Number.isFinite(result[key]))) {
    throw new Error('nonfinite FP64 comparison metric');
  }
  return result;
}

export function passesGate(metrics) {
  if (!metrics || !FIELDS.every(key => typeof metrics[key] === 'number'
    && Number.isFinite(metrics[key]))) return false;
  // A sqrt-product quotient may round one representable step above one.
  // This is a metric-domain guard, not a relaxation of any acceptance gate.
  if (Math.abs(metrics.cosine) > 1 + Number.EPSILON
    || Math.abs(metrics.worst_row_cosine) > 1 + Number.EPSILON
    || metrics.relative_l2 < 0 || metrics.max_abs_error < 0
    || metrics.exact_fraction < 0 || metrics.exact_fraction > 1) return false;
  return metrics.cosine >= THRESHOLDS.cosine_min
    && metrics.worst_row_cosine >= THRESHOLDS.worst_row_min
    && metrics.relative_l2 <= THRESHOLDS.relative_l2_max;
}

File: bench/deepseek-vision-gemmex/src/metrics.rs
// SPDX-License-Identifier: AGPL-3.0-only
//! Fixed fc1 schema and the pinned teacher's FP64 metric equations. No gates
//! or tolerance knobs here: candidate acceptance remains exact payload SHA.
use anyhow::{Result, ensure};
use deepseek_vision_p1::contract::check_bf16;
use serde_json::{Value, json};

const ROWS: usize = 20;
const COLS: usize = 5632;
#[derive(Default)]
struct Sum {
    value: f64,
    correction: f64,
}
impl Sum {
    fn add(&mut self, value: f64) {
        let next = self.value + value;
        self.correction += if self.value.abs() >= value.abs() {
            (self.value - next) + value
        } else {
            (value - next) + self.value
        };
        self.value = next;
    }
    fn total(&self) -> f64 {
        self.value + self.correction
    }
}
fn decode(raw: &[u8], index: usize) -> f64 {
    let bits = u16::from_le_bytes([raw[2 * index], raw[2 * index + 1]]);
    f32::from_bits(u32::from(bits) << 16) as f64
}
pub fn compare(actual: &[u8], reference: &[u8]) -> Result<Value> {
    ensure!(
        actual.len() == ROWS * COLS * 2 && reference.len() == actual.len(),
        "exact 20x5632 BF16 extent required"
    );
    check_bf16(actual)?;
    check_bf16(reference)?;
    let (mut aa, mut bb, mut dot, mut error2) = (
        Sum::default(),
        Sum::default(),
        Sum::default(),
        Sum::default(),
    );
    let mut worst = f64::INFINITY;
    let mut max_error = 0.0f64;
    let mut equal = 0usize;
    for row in 0..ROWS {
        let (mut ra, mut rb, mut rd) = (Sum::default(), Sum::default(), Sum::default());
        for col in 0..COLS {
            let index = row * COLS + col;
            let (a, b) = (decode(actual, index), decode(reference, index));
            ra.add(a * a);
            rb.add(b * b);
            rd.add(a * b);
            let error = a - b;
            error2.add(error * error);
            max_error = max_error.max(error.abs());
            if a == b {
                equal += 1;
            } // Torch numeric equality includes +0/-0.
        }
        let (a, b, d) = (ra.total(), rb.total(), rd.total());
        ensure!(a > 0.0 && b > 0.0, "zero norm comparison row {row}");
        worst = worst.min(d / (a.sqrt() * b.sqrt()));
        aa.add(a);
        bb.add(b);
        dot.add(d);
    }
    let reference_norm = bb.total().sqrt();
    let cosine = dot.total() / (aa.total().sqrt() * reference_norm);
    let relative_l2 = error2.total().sqrt() / reference_norm;
    let exact = (equal as f32 / (ROWS * COLS) as f32) as f64;
    ensure!(
        [cosine, worst, relative_l2, max_error, exact]
            .iter()
            .all(|v| v.is_finite()),
        "nonfinite FP64 metric"
    );
    Ok(
        json!({"cosine":cosine,"worst_row_cosine":worst,"relative_l2":relative_l2,
        "max_abs_error":max_error,"f32_exact_fraction":exact,
        "shape":[ROWS,COLS],"dtype":"BF16","reduction":"compensated serial FP64; Torch parallel FP64 can differ in final bits"}),
    )
}
